src/multi.c

```diff
@@ -296,12 +295,12 @@ void watchForKey(client *c, robj *key) {
         dictAdd(c->db->watched_keys,key,clients);
         incrRefCount(key);
     }
+    expireIfNeeded(c->db, key, 0);
```

Comment:
Deleting expired keys on WATCH makes making WATCH a command with side effects that need replication. Replicas delete expired keys only when their master deletes them. I'm not sure if we have a test that checks that WATCH doesn't need replication.

The reason we have an `expired` bit in watchedKey struct is described in this PR: #10256

Reply:
Didn't notice the history. I see, in replicas, `expireIfNeeded` not really delete the expired keys and this PR might not work (though I don't get the scenario why using `watch` in read-only replicas).

Reply:
When using WATCH in master, if an expired key is deleted, it needs to be propagated by sending DEL to the replicas, so WATCH needs to be marked as a 'write' command. That's why we didn't want WATCH to delete expired keys.

Reply:
Every `read` command also calls `expireIfNeeded`, this doesn't make it a `write` command. I think deleting expired keys makes no difference to clients, this can happend in any time any where, it's a hidden operation to clients.

Reply:
@oranagra do you want to bring clarity on this topic?

Reply:
What I can tell from the old discussions, the design with the `expired` bit in watchedKey was chosen because it solves the problem also in readonly replicas. It doesn't depend on whether expireIfNeeded actually deletes the expired key or not.

In read commands, the return value of expireIfNeeded is used and it doesn't matter if the key is deleted or not. In WATCH in this PR, only the side effect of deleting the key is used, which can't be trusted.

Not only readonly replicas have this problem. If clients are paused, expireIfNeeded also returns 1 without deleting the key.

Reply:
Seems to be more complex than I thought, I will close this PR. Thanks for the information and review.